src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,298 @@
+                    if (ifBodyEndPointIsReachable)
+                    {
+                        return InvertIfStyle.MoveIfBodyToElseClause;
+                    }
+                    else
+                    {
+                        if (SingleIfBodyStatement(ifStatement) &&
+                            SubsequentStatementsAreInTheSameBlock(ifStatement) &&
+                            ifBodySingleExitPointOpt?.RawKind == GetNearmostParentJumpStatementRawKind(ifStatement))
+                        {
+                            return InvertIfStyle.MoveSubsequenceToElseBody;
+                        }
+                        else
+                        {
+                            return InvertIfStyle.WithElseClause;
+                        }
+                    }
+                }
+                else
+                {
+                    if (ifBodyEndPointIsReachable)
+                    {
+                        if (subsequenceSingleExitPointOpt != null &&
+                            SingleSubsequenceStatement(ifStatement))
+                        {
+                            return InvertIfStyle.WithSubsequenceExitPoint;
+                        }
+                        else
+                        {
+                            return InvertIfStyle.MoveIfBodyToElseClause;
+                        }
+                    }
+                    else
+                    {
+                        if (SubsequentStatementsAreInTheSameBlock(ifStatement))
+                        {
+                            return InvertIfStyle.SwapIfBodyWithSubsequence;
+                        }
+                        else
+                        {
+                            return InvertIfStyle.MoveIfBodyToElseClause;
+                        }
+                    }
+                }
+            }
+
+            private bool IsEmptySubsequence(TIfStatementSyntax ifStatement)
+            {
+                foreach (var range in GetSubsequentStatementRange(ifStatement))
+                {
+                    if (!IsEmptyStatementRange(range))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private bool IsEmptyIfBody(TIfStatementSyntax ifStatement)
+            {
+                return IsEmptyStatementRange(GetIfBodyStatementRange(ifStatement));
+            }
+
+            private bool SingleIfBodyStatement(TIfStatementSyntax ifStatement)
+            {
+                var range = GetIfBodyStatementRange(ifStatement);
+                return range.first == range.last;
+            }
+
+            private bool SingleSubsequenceStatement(TIfStatementSyntax ifStatement)
+            {
+                using (var e = GetSubsequentStatementRange(ifStatement).GetEnumerator())
+                {
+                    return e.MoveNext() && e.Current.first == e.Current.last && !e.MoveNext();
+                }
```

Comment:
this is incredibly confusing :)

Reply:
what else could do here?

Reply:
```c#
var range = GetSubsequentStatementRange(ifStatement);
return range.Length == 1 && range[0].first == range[0].last;
```

or something thereabouts?

Reply:
1.  why does it need to eb an iterator?
2. an iterator can be realized.  It's not like this will end up being expensive :)